src/cargo/ops/cargo_package.rs

```diff
@@ -897,6 +903,36 @@ fn check_repo_state(
         Ok(dirty_files)
     }
 
+    /// Checks whether source files are symlinks and have been modified.
+    ///
+    /// This is required because those paths may link to a file outside the
+    /// current package root, but still under the git workdir, affecting the
+    /// final packaged `.crate` file.
+    fn dirty_symlinks(
+        pkg: &Package,
+        repo: &git2::Repository,
+        src_files: &[PathEntry],
+    ) -> CargoResult<HashSet<PathBuf>> {
+        let workdir = repo.workdir().unwrap();
+        let mut dirty_symlinks = HashSet::new();
+        for rel_path in src_files
+            .iter()
+            .filter(|p| p.is_symlink())
+            .map(|p| p.as_path())
+            // If inside package root. Don't bother checking git status.
+            .filter(|p| paths::strip_prefix_canonical(*p, pkg.root()).is_err())
+            // Handle files outside package root but under git workdir,
+            .filter_map(|p| paths::strip_prefix_canonical(p, workdir).ok())
+        {
+            // TODO: Should we warn users if there are like thousands of symlinks,
+            // which may hurt the performance?
+            if repo.status_file(&rel_path)? != git2::Status::CURRENT {
+                dirty_symlinks.insert(workdir.join(rel_path));
```

Comment:
We could potentially cache git status for each repo then. That leaves to future improvement if this one has got merged.

Reply:
See <https://github.com/rust-lang/cargo/pull/14985>.